docs/site/Loopback-component-authorization.md

```diff
@@ -8,543 +8,377 @@ permalink: /doc/en/lb4/Loopback-component-authorization.html
+  @authorize({resource: 'order', scopes: ['patch'], voters: [compareId]})
+  async patchOrders(
+    @param.path.string('userId') userId: string,
+    @requestBody() order: Partial<Order>,
+    @param.query.string('where') where?: Where<Order>,
+  ): Promise<Count> {
+    return this.userRepo.orders(userId).patch(order, where);
   }
-}
 ```
 
-- Add user level permissions array to the user model. Do this if there is a use
-  case of explicit allow/deny of permissions at user-level in the application.
-  You can skip otherwise.
-
-{% include code-caption.html content="/src/models/user.model.ts" %}
+In the above example `compareId()` is a authorizing function which is provided
+as a voter in the decorator for the `patchOrders()` method.
+
+## Authorization by decision matrix
+
+The final decision to allow access for a subject is done by the interceptor by
+creating a decision matrix from the voting results (from all the `authorizer` or
+`voter` functions of an endpoint).
+
+The following table illustrates an example decision matrix with 3 votes and
+corresponding options.
+
+| Vote #1 | Vote # 2 | Vote #3 | Options                  | Final Decision |
+| ------- | -------- | ------- | ------------------------ | -------------- |
+| Deny    | Deny     | Deny    | **any**                  | Deny           |
+| Allow   | Allow    | Allow   | **any**                  | Allow          |
+| Abstain | Allow    | Abstain | **any**                  | Allow          |
+| Abstain | Deny     | Abstain | **any**                  | Deny           |
+| Deny    | Allow    | Abstain | {precedence: Deny}       | Deny           |
+| Deny    | Allow    | Abstain | {precedence: Allow}      | Allow          |
+| Allow   | Abstain  | Deny    | {precedence: Deny}       | Deny           |
+| Allow   | Abstain  | Deny    | {precedence: Allow}      | Allow          |
+| Abstain | Abstain  | Abstain | {defaultDecision: Deny}  | Deny           |
+| Abstain | Abstain  | Abstain | {defaultDecision: Allow} | Allow          |
+
+- Here, if suppose there is an `authorizer` function and 2 voters for an
+  endpoint.
+  - if the `authorizer` function returns `ALLOW`, but voter 1 in authorize
+    decorator returns `ABSTAIN` and voter 2 in decorator returns `DENY`.
+  - In this case, if the options provided while
+    [registering the authorization component](#Authorization-Component),
+    provides precedence as `DENY`, then the access for the subject is denied to
+    the endpoint.
+
+## Enforcer Libraries
+
+Enforcer libraries help developers in coding security policies as configurations
+and provides out-of-the-box matching rules, strategies and authorization
+patterns. These libraries also help with mundane tasks like mapping user
+profiles to scopes and roles, modifying configurations dynamically, etc. Please
+look at the
+[loopback shopping example](https://github.com/strongloop/loopback4-example-shopping)
+to see how CasBin library is injected as an enforcer into authorization
```

Comment:
> Please look at the loopback shopping example to see how CasBin library is injected as an enforcer into authorization provider.

Please look at the loopback shopping example to see how CasBin library is injected as an enforcer into the authorization provider.

Reply:
agreed